File: rs/moq-lite/src/lite/subscribe.rs
use std::borrow::Cow;

use crate::{
	Path,
	coding::{Decode, DecodeError, Encode},
	lite::{Message, Version},
};

/// Sent by the subscriber to request all future objects for the given track.
///
/// Objects will use the provided ID instead of the full track name, to save bytes.
#[derive(Clone, Debug)]
pub struct Subscribe<'a> {
	pub id: u64,
	pub broadcast: Path<'a>,
	pub track: Cow<'a, str>,
	pub priority: u8,
	pub ordered: bool,
	pub max_latency: std::time::Duration,
	pub start_group: Option<u64>,
	pub end_group: Option<u64>,
}

impl Message for Subscribe<'_> {
	fn decode_msg<R: bytes::Buf>(r: &mut R, version: Version) -> Result<Self, DecodeError> {
		let id = u64::decode(r, version)?;
		let broadcast = Path::decode(r, version)?;
		let track = Cow::<str>::decode(r, version)?;
		let priority = u8::decode(r, version)?;

		let (ordered, max_latency, start_group, end_group) = match version {
			Version::Draft03 => {
				let ordered = u8::decode(r, version)? != 0;
				let max_latency = std::time::Duration::decode(r, version)?;
				let start_group = Option::<u64>::decode(r, version)?;
				let end_group = Option::<u64>::decode(r, version)?;
				(ordered, max_latency, start_group, end_group)
			}
			Version::Draft01 | Version::Draft02 => (false, std::time::Duration::ZERO, None, None),
		};

		Ok(Self {
			id,
			broadcast,
			track,
			priority,
			ordered,
			max_latency,
			start_group,
			end_group,
		})
	}

	fn encode_msg<W: bytes::BufMut>(&self, w: &mut W, version: Version) {
		self.id.encode(w, version);
		self.broadcast.encode(w, version);
		self.track.encode(w, version);
		self.priority.encode(w, version);

		match version {
			Version::Draft03 => {
				(self.ordered as u8).encode(w, version);
				self.max_latency.encode(w, version);
				self.start_group.encode(w, version);
				self.end_group.encode(w, version);
			}
			Version::Draft01 | Version::Draft02 => {}
		}
	}
}

#[derive(Clone, Debug)]
pub struct SubscribeOk {
	pub priority: u8,
	pub ordered: bool,
	pub max_latency: std::time::Duration,
	pub start_group: Option<u64>,
	pub end_group: Option<u64>,
}

impl Message for SubscribeOk {
	fn encode_msg<W: bytes::BufMut>(&self, w: &mut W, version: Version) {
		match version {
			Version::Draft03 => {
				self.priority.encode(w, version);
				(self.ordered as u8).encode(w, version);
				self.max_latency.encode(w, version);
				self.start_group.encode(w, version);
				self.end_group.encode(w, version);
			}
			Version::Draft01 => {
				self.priority.encode(w, version);
			}
			Version::Draft02 => {}
		}
	}

	fn decode_msg<R: bytes::Buf>(r: &mut R, version: Version) -> Result<Self, DecodeError> {
		match version {
			Version::Draft03 => {
				let priority = u8::decode(r, version)?;
				let ordered = u8::decode(r, version)? != 0;
				let max_latency = std::time::Duration::decode(r, version)?;
				let start_group = Option::<u64>::decode(r, version)?;
				let end_group = Option::<u64>::decode(r, version)?;

				Ok(Self {
					priority,
					ordered,
					max_latency,
					start_group,
					end_group,
				})
			}
			Version::Draft01 => Ok(Self {
				priority: u8::decode(r, version)?,
				ordered: false,
				max_latency: std::time::Duration::ZERO,
				start_group: None,
				end_group: None,
			}),
			Version::Draft02 => Ok(Self {
				priority: 0,
				ordered: false,
				max_latency: std::time::Duration::ZERO,
				start_group: None,
				end_group: None,
			}),
		}
	}
}

/// Sent by the subscriber to update subscription parameters.
///
/// Draft03 only.
#[allow(dead_code)]
#[derive(Clone, Debug)]
pub struct SubscribeUpdate {
	pub priority: u8,
	pub ordered: bool,
	pub max_latency: std::time::Duration,
	pub start_group: Option<u64>,
	pub end_group: Option<u64>,
}

impl Message for SubscribeUpdate {
	fn decode_msg<R: bytes::Buf>(r: &mut R, version: Version) -> Result<Self, DecodeError> {
		match version {
			Version::Draft01 | Version::Draft02 => {
				unreachable!("subscribe update not supported for version: {:?}", version);
			}
			Version::Draft03 => {}
		}

		let priority = u8::decode(r, version)?;
		let ordered = u8::decode(r, version)? != 0;
		let max_latency = std::time::Duration::decode(r, version)?;
		let start_group = match u64::decode(r, version)? {
			0 => None,
			group => Some(group - 1),
		};
		let end_group = match u64::decode(r, version)? {
			0 => None,
			group => Some(group - 1),
		};

		Ok(Self {
			priority,
			ordered,
			max_latency,
			start_group,
			end_group,
		})
	}

	fn encode_msg<W: bytes::BufMut>(&self, w: &mut W, version: Version) {
		match version {
			Version::Draft01 | Version::Draft02 => {
				unreachable!("subscribe update not supported for version: {:?}", version);
			}
			Version::Draft03 => {}
		}

		self.priority.encode(w, version);
		(self.ordered as u8).encode(w, version);
		self.max_latency.encode(w, version);

		match self.start_group {
			Some(start_group) => (start_group + 1).encode(w, version),
			None => 0u64.encode(w, version),
		}

		match self.end_group {
			Some(end_group) => (end_group + 1).encode(w, version),
			None => 0u64.encode(w, version),
		}
	}
}

File: js/lite/src/lite/subscribe.ts
import * as Path from "../path.ts";
import type { Reader, Writer } from "../stream.ts";
import * as Message from "./message.ts";
import { Version } from "./version.ts";

export class SubscribeUpdate {
	priority: number;
	ordered: boolean;
	maxLatency: number;
	startGroup: number;
	endGroup: number;

	constructor(
		priority: number,
		ordered: boolean = true,
		maxLatency: number = 0,
		startGroup: number = 0,
		endGroup: number = 0,
	) {
		this.priority = priority;
		this.ordered = ordered;
		this.maxLatency = maxLatency;
		this.startGroup = startGroup;
		this.endGroup = endGroup;
	}

	async #encode(w: Writer, version?: Version) {
		await w.u8(this.priority);
		if (version === Version.DRAFT_03) {
			await w.bool(this.ordered);
			await w.u53(this.maxLatency);
			await w.u53(this.startGroup);
			await w.u53(this.endGroup);
		}
	}

	static async #decode(r: Reader, version?: Version): Promise<SubscribeUpdate> {
		const priority = await r.u8();
		if (version === Version.DRAFT_03) {
			const ordered = await r.bool();
			const maxLatency = await r.u53();
			const startGroup = await r.u53();
			const endGroup = await r.u53();
			return new SubscribeUpdate(priority, ordered, maxLatency, startGroup, endGroup);
		}
		return new SubscribeUpdate(priority);
	}

	async encode(w: Writer, version?: Version): Promise<void> {
		return Message.encode(w, (w) => this.#encode(w, version));
	}

	static async decode(r: Reader, version?: Version): Promise<SubscribeUpdate> {
		return Message.decode(r, (r) => SubscribeUpdate.#decode(r, version));
	}

	static async decodeMaybe(r: Reader, version?: Version): Promise<SubscribeUpdate | undefined> {
		return Message.decodeMaybe(r, (r) => SubscribeUpdate.#decode(r, version));
	}
}

export class Subscribe {
	id: bigint;
	broadcast: Path.Valid;
	track: string;
	priority: number;
	ordered: boolean;
	maxLatency: number;
	startGroup: number;
	endGroup: number;

	constructor(
		id: bigint,
		broadcast: Path.Valid,
		track: string,
		priority: number,
		ordered: boolean = true,
		maxLatency: number = 0,
		startGroup: number = 0,
		endGroup: number = 0,
	) {
		this.id = id;
		this.broadcast = broadcast;
		this.track = track;
		this.priority = priority;
		this.ordered = ordered;
		this.maxLatency = maxLatency;
		this.startGroup = startGroup;
		this.endGroup = endGroup;
	}

	async #encode(w: Writer, version?: Version) {
		await w.u62(this.id);
		await w.string(this.broadcast);
		await w.string(this.track);
		await w.u8(this.priority);
		if (version === Version.DRAFT_03) {
			await w.bool(this.ordered);
			await w.u53(this.maxLatency);
			await w.u53(this.startGroup);
			await w.u53(this.endGroup);
		}
	}

	static async #decode(r: Reader, version?: Version): Promise<Subscribe> {
		const id = await r.u62();
		const broadcast = Path.from(await r.string());
		const track = await r.string();
		const priority = await r.u8();
		if (version === Version.DRAFT_03) {
			const ordered = await r.bool();
			const maxLatency = await r.u53();
			const startGroup = await r.u53();
			const endGroup = await r.u53();
			return new Subscribe(id, broadcast, track, priority, ordered, maxLatency, startGroup, endGroup);
		}
		return new Subscribe(id, broadcast, track, priority);
	}

	async encode(w: Writer, version?: Version): Promise<void> {
		return Message.encode(w, (w) => this.#encode(w, version));
	}

	static async decode(r: Reader, version?: Version): Promise<Subscribe> {
		return Message.decode(r, (r) => Subscribe.#decode(r, version));
	}
}

export class SubscribeOk {
	// The version
	readonly version: Version;
	priority?: number;
	ordered?: boolean;
	maxLatency?: number;
	startGroup?: number;
	endGroup?: number;

	constructor({
		version,
		priority = undefined,
		ordered = undefined,
		maxLatency = undefined,
		startGroup = undefined,
		endGroup = undefined,
	}: {
		version: Version;
		priority?: number;
		ordered?: boolean;
		maxLatency?: number;
		startGroup?: number;
		endGroup?: number;
	}) {
		this.version = version;
		this.priority = priority;
		this.ordered = ordered;
		this.maxLatency = maxLatency;
		this.startGroup = startGroup;
		this.endGroup = endGroup;
	}

	async #encode(w: Writer) {
		if (this.version === Version.DRAFT_03) {
			await w.u8(this.priority ?? 0);
			await w.bool(this.ordered ?? true);
			await w.u53(this.maxLatency ?? 0);
			await w.u53(this.startGroup ?? 0);
			await w.u53(this.endGroup ?? 0);
		} else if (this.version === Version.DRAFT_02) {
			// noop
		} else if (this.version === Version.DRAFT_01) {
			await w.u8(this.priority ?? 0);
		}
	}

	static async #decode(version: Version, r: Reader): Promise<SubscribeOk> {
		let priority: number | undefined;
		let ordered: boolean | undefined;
		let maxLatency: number | undefined;
		let startGroup: number | undefined;
		let endGroup: number | undefined;

		if (version === Version.DRAFT_03) {
			priority = await r.u8();
			ordered = await r.bool();
			maxLatency = await r.u53();
			startGroup = await r.u53();
			endGroup = await r.u53();
		} else if (version === Version.DRAFT_02) {
			// noop
		} else if (version === Version.DRAFT_01) {
			priority = await r.u8();
		}

		return new SubscribeOk({ version, priority, ordered, maxLatency, startGroup, endGroup });
	}

	async encode(w: Writer): Promise<void> {
		return Message.encode(w, this.#encode.bind(this));
	}

	static async decode(r: Reader, version: Version): Promise<SubscribeOk> {
		return Message.decode(r, SubscribeOk.#decode.bind(SubscribeOk, version));
	}
}
